### DISCUSSION TITLE
pursue better contrast on outlines for menus, dialogs, etc. especially in dark mode

### DISCUSSION THREAD

**Discussion by mbgower** [carbon-design-system/carbon#12957] [Accessibility]
Especially in dark mode, the outlines of tooltips, menus, dialogs and other types of overlays can be under-pronounced.
In the below screenshot of a dialog, the low constrast between the gray box and the darker gray background results in an experience where things like the X close buttons seems to be floating in the middle of the page.
![image](https://user-images.githubusercontent.com/14298245/119692927-af953200-be00-11eb-95ea-4aee079a0547.png)

![image](https://user-images.githubusercontent.com/14298245/119890781-18f26f00-beed-11eb-9dd3-1e5b5fb3db40.png)

Although 'white space' can be used to separate the text of such non-persistent content from the background a demarcation of the border that achieves 3:1 would reduce potential failures of Non-text Contrast.

Note that this is more likely in the realm of 'ehancement' than 'bug' -- I don't think in most cases we'd fail [Non-text Contrast](https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html), because one can argue that you don't need the border to identify the component. But the benefit of more contrast is clear and backed up by our research. This is especially useful for low vision users, who may have more difficulty locating this material when it appears, as well as more difficulty differentiating between this material and content behind it.

**Comment by mbgower** [comment]
One interesting thing I noticed was that in the [Overflow menu on the usage page](https://www.carbondesignsystem.com/components/overflow-menu/usage/), the focus indicator is used to create that clear boundary.
That's a little unusual, but may be interesting to explore as an overall solution.
![image](https://user-images.githubusercontent.com/14298245/119890946-45a68680-beed-11eb-931d-bc767283a1a0.png)

**Comment by joshblack** [comment]
cc @aagonzales @laurenmrice

**Comment by aagonzales** [comment]
@mbgower That outline is what actually happens on all our dropdowns. I think a lot of our illustrated dropdowns in website images aren't showing the focus border but it does always appear in code for dropdowns. The focus stays on the whole menu then the hover style is what appears when arrowing through the options. It gets to be a lot of you have a focus outline around the menu and also a focus outline around the items. 

https://react.carbondesignsystem.com/?path=/story/components-dropdown--default
![image](https://user-images.githubusercontent.com/11670886/120014015-266c2f80-bfa7-11eb-890d-7566f9ba9e04.png)

**Comment by mbgower** [comment]
@aagonzales 
> That outline is what actually happens on all our dropdowns...The focus stays on the whole menu then the hover style is what appears when arrowing through the options. 

I think this can be a good approach where the component takes some kind of outline focus around the whole. Please note that the hover style shading when arrowing through the options needs to meet 3:1 against the non-focused items!

However, this approach doesn't address the problem for something like a dialog (where the focus would normally only be on the elements within it that take focus). So I think the request to improve contrast on outlines in such situations still exists?

**Comment by sstrubberg** [comment]
This might need to turn into a Github Discussion rather than an issue that needs to be addressed.

**Comment by mbgower** [comment]
I also believe that some of the work @thyhmdo has done on Tiles may be appropriate to consider here.
Thy, in regard to outline, do you see any of the considerations you had around tile contrast enhancements being relevant to increasing the outline of dialogs and complex components that do not take focus around the whole?

**Comment by thyhmdo** [comment]
Hey Michael, there are couple thoughts. 

1. I think we should look at more examples of what these components look like on the product pages. For example, when I first come to this Security page. It is an issue that the modal looks like it's blending in with the background even though there is an overlay background. 

![image](https://user-images.githubusercontent.com/77697351/211844938-23094bd9-6a88-4058-8e94-f45d2d07ec33.png)


2. However, there are also a case where they use a lighter background modal and that looks much better. I wonder if these cases have enough contrast for accessibility? If these pass the accessibility requirements, then we can advise people to avoid using the modals those look like it could create the issue here. 

![image](https://user-images.githubusercontent.com/77697351/211846172-04f9359d-2309-4440-9cdd-e4dcfaa8c23a.png)

![image](https://user-images.githubusercontent.com/77697351/211846475-d02ce4bf-66ae-4dde-8e9c-1f99b619171b.png)


3. But here is just the modal case. Seems like the dropdown focus line could work. Are there any other specific components you think of for this issue? The header menu could be a problem? While Security has a subtle border around it, Cloud doesn't have it. Note that our current menu doesn't have a border either. 

![image](https://user-images.githubusercontent.com/77697351/211848342-41bccd59-3bc7-46ce-b9ce-ef52b57e86cc.png)
![image](https://user-images.githubusercontent.com/77697351/211848847-e18348d0-263f-4362-b354-ab393ba8e7b2.png)

**Comment by mbgower** [comment]
thanks, Thy. I think the outline around that Dallas menu/dropdown is working okay. They've basically stuck with the black of the top nav bar for the options list background, and then used grey 80 both for the item with focus and as the stroke on the outline. Ironically it is not very high contrast either. If they used a gray 70 that could _almost_ get to 3:1 while letting the the Gray 10 text still contrast more than 4.5:1 when the item is selected (Dallas). It's actually going to be hard to indicate the option with focus using just a fill because it's tough to get a color that both contrasts 3:1 against the fill of unselected options and also 4.5:1 between the fill and the text colour.

A safer option is to put a focus indicator around the menu/list as well as the item. It wouldn't even need to be as thick as the one around Manage in the bottom example, but a 1 pixel around the whole list would certainly pump it up and ensure that when the focus appearance comes into force with WCAG 2.2, this would pass. (As it is, it will fail the contrast on the option with focus.)

So I guess there are two different topics we should be careful not to mix up too much:
1. a better distinction of the border of dialogs and dropdowns/menus (nice to have)
2. a 3:1 distinction between an item/option with focus and one without focus (arguably necessary now, and definitely a requirement in WCAG 2.2)

**Comment by laurenmrice** [comment]
Another instance from the zAIOps team (Example provided by Phoebe Cai) where dropdown menus overlay dashboard content using similar colors. I do know currently that the menu would technically get a focus outline here, but we have had recent discussions on if we need to change it. 

![image (36)](https://user-images.githubusercontent.com/43969356/232773355-5de8f64a-0ec6-4353-8e9f-94953ee38631.png)

**Reply by Phoebecdy91** [reply]
For some of the dropdown, like Combo box & Filterable multiselect, the dropdown menu doesn't have a focus border. But for the default and multi-select one, the menu has the focus border. I would suggest to keep consistent. 
And I will put one of our dev env for implementing this. Seems the focus border makes it too different from others.
<img width="1531" alt="图片" src="https://user-images.githubusercontent.com/23204849/232945873-206740fc-432a-40c5-a058-fa1e28f429d6.png">

**Comment by mbgower** [comment]
Thanks, @laurenmrice. I will note that WCAG 2.2's Focus Appearance has been moved to a AAA. It's not that people did not want to have a requirement for enough contrast for the item with the focus, but we could not find a way of consistently testing it that did not get very tricky on variable backgrounds.
The non-text contrast requirements remain for Focus, but there's no requirement that it necessarily contrast with the non-contrasted item. In my opinion, Carbon should still try to make the delineation on dropdowns in particular better, but the existing contrast requirement is that the indicator for which item in the dropdown is currently focused needs to be at least 3:1, which it isn't using just the hover effect.
